impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -101,76 +95,90 @@ void FenceWaiterVK::Main() {
     // critical section. Copy the array of entries and immediately unlock the
     // mutex.
     WaitSet wait_set = wait_set_;
-
-    const auto terminate = terminate_;
-
     lock.unlock();
 
+    const auto terminate = terminate_;
     if (terminate) {
+      WaitUntilEmpty();
       break;
     }
 
-    // Check if the context had died in the meantime.
-    auto device_holder = device_holder_.lock();
-    if (!device_holder) {
+    if (!Wait(wait_set)) {
       break;
     }
+  }
+}
 
-    const auto& device = device_holder->GetDevice();
+void FenceWaiterVK::WaitUntilEmpty() {
+  while (!wait_set_.empty() && Wait(wait_set_)) {
```

Comment:
Maybe add a note that this doesn't lock because its only called when the threads are joined?

Reply:
Or does it?

Reply:
It doesn't lock because (as-per your suggestion on AddFence), it's not possible to mutate wait_set_ once terminate_ is set.